fix: watch buffers instead of editors

aleclarson opened the pull request:
Watching `TextBuffer` objects instead of `TextEditor` objects worksbetter, because editors of the same file share the same buffer.
Also, the `onDidDestroy` listener was broken, because it didn't look for editors of the same path before mutating the `watching` object. The first solution I thought of was to keep track of how many editors are open per watched file, but that also meant attaching `onDidDestroy` listeners to each editor. I found it much simpler to attach listeners to the shared TextBuffer objects.

The package of each watched file has an `onDidDeactivate` listener attached. This allows us to stop watching the files of that package when/if the grammar is deactivated. This listener is reattached whenever the grammar is reloaded, because the `Package` object is recreated.

When a watched file is deleted, we stop watching it.

When a watched file is renamed, we stop watching it *if* the new file path does not match the `grammarRE` pattern. We assume its package is still the same, but maybe this isn't a safe assumption?

We *don't* yet listen for renamed package of watched files, but this could be a nice addition sometime in the future.

The buffer listeners are now properly disposed of when live reloading is disabled by the user.

The path of each `Package` is already resolved when it's a symlink, so `getPackageNameByFilePath` has been refactored and renamed to `resolvePackageName`. The function still looks at the `path` of
every active package until it finds a match, which I agree is better than assuming the package directory's basename is equal to its "name" field in `package.json`.

aleclarson committed:
fix: watch buffers instead of editors

Watching `TextBuffer` objects instead of `TextEditor` objects works
better, because editors of the same file share the same buffer.
Also, the `onDidDestroy` listener was broken, because it didn't
look for editors of the same path before mutating the `watching`
object. The first solution I thought of was to keep track of how
many editors are open per watched file, but that also meant
attaching `onDidDestroy` listeners to each editor. I found it much
simpler to attach listeners to the shared TextBuffer objects.

The package of each watched file has an `onDidDeactivate`
listener attached. This allows us to stop watching the files
of that package when/if the grammar is deactivated. This listener
is reattached whenever the grammar is reloaded, because the `Package`
object is recreated.

When a watched file is deleted, we stop watching it.

When a watched file is renamed, we stop watching it *if* the new file
path does not match the `grammarRE` pattern. We assume its package
is still the same, but maybe this isn't a safe assumption?

We *don't* yet listen for renamed package of watched files, but
this could be a nice addition sometime in the future.

The buffer listeners are now properly disposed of when live reloading
is disabled by the user.

The `grammarRE` string now has its '/' characters properly escaped.

ldez commented on lib/main.coffee:
```diff
@@ -37,47 +28,79 @@ module.exports =
+        @buffers.clear()
+        return
+
+      resolvePackageName = (grammarPath) ->
```
extract the function

ldez commented on lib/main.coffee:
```diff
@@ -37,47 +28,79 @@ module.exports =
+        for pack in atom.packages.getActivePackages()
+          return pack.name if pack.path is packPath
+
+      isGrammarPath = (filePath) ->
```
extract the function

aleclarson replied:
It's not used outside that scope, though. It's better to keep it close to where it is used, because minor perf improvements to config changes are not a necessity.

ldez commented on lib/main.coffee:
```diff
@@ -37,47 +28,79 @@ module.exports =
+      resolvePackageName = (grammarPath) ->
+        packPath = path.resolve grammarPath, '../..'
+        for pack in atom.packages.getActivePackages()
+          return pack.name if pack.path is packPath
```
```coffee
 packPath = if stats.isSymbolicLink then fs.realpathSync pack.path else pack.path
```

aleclarson replied:
👍

aleclarson replied:
`isSymbolicLink` is a function, not a boolean

ldez commented on lib/main.coffee:
```diff
@@ -96,12 +119,11 @@ module.exports =
-            if grammar.packageName is packName
-              debug and console.log 'Updating grammar for editor: ', editor
-              editor.setGrammar grammars[grammar.scopeName]
+        atom.workspace.getTextEditors().forEach (editor) ->
```
I don't agree with this change

aleclarson replied:
It's a nitpick either way, so you get precedence here.

I'll change it back to a `for` loop, but the `do (editor) ->` part is unnecessary.

ldez commented:
Your assumption about `getPackageNameByFilePath` and symlink is false.
On Linux, `pack.path` contains symlink (`apm link -d`).

aleclarson commented:
> Your assumption about getPackageNameByFilePath and symlink is false.

Good to know, thanks!

Doesn't `rootPath` need to be resolved, too? If `Package` paths are not always the real path, can we really assume the `TextBuffer` path is the real path?

ldez commented on lib/main.coffee:
```diff
@@ -37,48 +28,87 @@ module.exports =
+        @buffers.clear()
+        return
+
+      resolveSymlink = (filePath) ->
```
please extract the functions

aleclarson replied:
It's not used outside that scope. Isn't it better to keep it close to where it is used, since minor perf improvements to config changes are not a necessity?

ldez replied:
please extract the functions.

ldez commented:
please do commits instead of amend/rebase, it's more easy to review the changes.

aleclarson commented:
ok sorry

- The `grammarRE` string now has its `/` chars properly escaped
- Switched back to `for` loop for `getTextEditors()`
- Added `resolveSymlink`

aleclarson committed:
hoist the helper functions

aleclarson committed:
add missing 'fs' import

aleclarson commented:
I will squash before merge

aleclarson committed:
fix: stop watching when all editors of a file are closed

aleclarson committed:
fix: add missing `isGrammarPath` call

aleclarson committed:
fix: protect against untitled editors

aleclarson commented:
I think I've squashed all the bugs. Let me know when to squash the commits. 🙂

aleclarson commented:
Great to see you got time to merge this. Cheers mate!
